src/jit/gentree.cpp

```diff
@@ -11812,6 +11812,42 @@ GenTree* Compiler::gtFoldExprCompare(GenTree* tree)
     return cons;
 }
 
+//------------------------------------------------------------------------
+// gtCreateHandleCompare: generate a type handle comparison
+//
+// Arguments:
+//    oper -- comparison operation (equal/not equal)
+//    op1 -- first operand
+//    op2 -- second operand
+//    canComparePointers -- flag indicating whether it's safe to compare handles as integers
+//
+// Returns:
+//    Type comparison tree
+//
+
+GenTree* Compiler::gtCreateHandleCompare(genTreeOps oper, GenTree* op1, GenTree* op2, bool canComparePointers)
+{
+    // If we can compare pointers directly, just emit the binary operation
+    if (canComparePointers)
+    {
+        return gtNewOperNode(oper, TYP_INT, op1, op2);
+    }
+    
+    // CoreRT has a helper to do comparison otherwise. This helper is not present in CoreCLR.
+    assert(IsTargetAbi(CORINFO_CORERT_ABI));
+
+    // Emit a call to a runtime helper
+    GenTreeArgList* helperArgs = gtNewArgList(op1, op2);
+    GenTree* ret = gtNewHelperCallNode(CORINFO_HELP_ARE_TYPES_EQUIVALENT, TYP_INT, helperArgs);
```

Comment:
This seems unlikely to work in all situations. I assume that there are cases where the compare is used by a JTRUE node, that one expects a compare node but now you're returning a CALL/NOT node. You should probably produce something like `EQ/NE(CALL, 0)`

Reply:
Agree.